crates/ibc/src/core/ics04_channel/events.rs

```diff
@@ -1,719 +1,733 @@
-                AbciEvent {
-                    type_str,
-                    attributes,
-                }
-            }
-        })+
-    };
-}
-
-impl_from_ibc_to_abci_event!(
-    OpenInit,
-    OpenTry,
-    OpenAck,
-    OpenConfirm,
-    CloseInit,
-    CloseConfirm
-);
-
-#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
-pub struct SendPacket {
-    pub packet: Packet,
+impl From<CloseConfirm> for AbciEvent {
+    fn from(o: CloseConfirm) -> Self {
+        AbciEvent {
+            type_str: IbcEventType::CloseConfirmChannel.as_str().to_string(),
+            attributes: vec![
+                o.port_id.into(),
+                o.channel_id.into(),
+                o.counterparty_port_id.into(),
+                o.counterparty_channel_id.into(),
+                o.connection_id.into(),
+            ],
+        }
+    }
 }
 
-impl SendPacket {
-    pub fn src_port_id(&self) -> &PortId {
-        &self.packet.source_port
+/// A `ChannelClosed` event is emitted when a channel is closed as a result of a packet timing out. Note that
+/// since optimistic packet sends (i.e. send a packet before channel handshake is complete) are supported,
+/// we might not have a counterparty channel id value yet. This would happen if a packet is sent right
+/// after a `ChannelOpenInit` message.
+#[derive(Debug)]
+pub struct ChannelClosed {
+    port_id: PortIdAttribute,
+    channel_id: ChannelIdAttribute,
+    counterparty_port_id: PortIdAttribute,
+    maybe_counterparty_channel_id: Option<ChannelIdAttribute>,
+    connection_id: ConnectionIdAttribute,
+    channel_ordering: ChannelOrderingAttribute,
+}
+
+impl ChannelClosed {
+    pub fn new(
+        port_id: PortId,
+        channel_id: ChannelId,
+        counterparty_port_id: PortId,
+        maybe_counterparty_channel_id: Option<ChannelId>,
+        connection_id: ConnectionId,
+        channel_ordering: Order,
+    ) -> Self {
+        Self {
+            port_id: port_id.into(),
+            channel_id: channel_id.into(),
+            counterparty_port_id: counterparty_port_id.into(),
+            maybe_counterparty_channel_id: maybe_counterparty_channel_id.map(|c| c.into()),
+            connection_id: connection_id.into(),
+            channel_ordering: channel_ordering.into(),
+        }
     }
-    pub fn src_channel_id(&self) -> &ChannelId {
-        &self.packet.source_channel
+    pub fn port_id(&self) -> &PortId {
+        &self.port_id.port_id
     }
-    pub fn dst_port_id(&self) -> &PortId {
-        &self.packet.destination_port
+    pub fn channel_id(&self) -> &ChannelId {
+        &self.channel_id.channel_id
     }
-    pub fn dst_channel_id(&self) -> &ChannelId {
-        &self.packet.destination_channel
+    pub fn counterparty_port_id(&self) -> &PortId {
+        &self.counterparty_port_id.port_id
+    }
+    pub fn counterparty_channel_id(&self) -> Option<ChannelId> {
+        self.maybe_counterparty_channel_id
```

Comment:
Nit: The clone can be avoided with an `impl AsRef<ChannelId> for ChannelIdAttribute` and ->
```rust
pub fn counterparty_channel_id(&self) -> Option<ChannelId> {        
    self.maybe_counterparty_channel_id.as_ref().map(|c| c.as_ref())
}
```

Reply:
[3ee4d71](https://github.com/cosmos/ibc-rs/pull/198/commits/3ee4d71b1a3445d3c5f94e3c657da18517267b13)